File: skills/math-expression-parser/reference/src/tokenizer.test.ts
import { describe, test, expect } from "bun:test";
import { tokenize } from "./tokenizer";

describe("tokenizer", () => {
  test("empty string", () => {
    expect(tokenize("")).toEqual([]);
  });

  test("whitespace only", () => {
    expect(tokenize("   \t\n\r  ")).toEqual([]);
  });

  test("single integer", () => {
    expect(tokenize("42")).toEqual([{ kind: "number", value: "42" }]);
  });

  test("decimal number", () => {
    expect(tokenize("3.14")).toEqual([{ kind: "number", value: "3.14" }]);
  });

  test("number starting with dot", () => {
    expect(tokenize(".5")).toEqual([{ kind: "number", value: ".5" }]);
  });

  test("all operators", () => {
    const tokens = tokenize("+ - * / % **");
    expect(tokens).toEqual([
      { kind: "plus", value: "+" },
      { kind: "minus", value: "-" },
      { kind: "star", value: "*" },
      { kind: "slash", value: "/" },
      { kind: "percent", value: "%" },
      { kind: "power", value: "**" },
    ]);
  });

  test("parentheses", () => {
    const tokens = tokenize("(1)");
    expect(tokens).toEqual([
      { kind: "lparen", value: "(" },
      { kind: "number", value: "1" },
      { kind: "rparen", value: ")" },
    ]);
  });

  test("complex expression", () => {
    const tokens = tokenize("2 + 3 * (4 - 1)");
    expect(tokens).toEqual([
      { kind: "number", value: "2" },
      { kind: "plus", value: "+" },
      { kind: "number", value: "3" },
      { kind: "star", value: "*" },
      { kind: "lparen", value: "(" },
      { kind: "number", value: "4" },
      { kind: "minus", value: "-" },
      { kind: "number", value: "1" },
      { kind: "rparen", value: ")" },
    ]);
  });

  test("power operator distinguished from multiply", () => {
    const tokens = tokenize("2**3*4");
    expect(tokens).toEqual([
      { kind: "number", value: "2" },
      { kind: "power", value: "**" },
      { kind: "number", value: "3" },
      { kind: "star", value: "*" },
      { kind: "number", value: "4" },
    ]);
  });

  test("no whitespace", () => {
    const tokens = tokenize("1+2");
    expect(tokens).toEqual([
      { kind: "number", value: "1" },
      { kind: "plus", value: "+" },
      { kind: "number", value: "2" },
    ]);
  });

  test("multiple decimals in one number throws", () => {
    expect(() => tokenize("1.2.3")).toThrow("Unexpected character '.'");
  });

  test("unrecognized character throws", () => {
    expect(() => tokenize("2 @ 3")).toThrow("Unexpected character '@'");
  });

  test("unrecognized character reports position", () => {
    expect(() => tokenize("2 @ 3")).toThrow("position 2");
  });
});

File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/test/kotlin/mathexpr/TokenizerTest.kt
package mathexpr

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class TokenizerTest {
    @Test
    fun `tokenizes empty string`() {
        assertEquals(emptyList(), tokenize(""))
    }

    @Test
    fun `tokenizes whitespace only`() {
        assertEquals(emptyList(), tokenize("   \t\n  "))
    }

    @Test
    fun `tokenizes number`() {
        val tokens = tokenize("42")
        assertEquals(1, tokens.size)
        assertEquals(TokenKind.NUMBER, tokens[0].kind)
        assertEquals("42", tokens[0].value)
    }

    @Test
    fun `tokenizes decimal number`() {
        val tokens = tokenize("3.14")
        assertEquals(1, tokens.size)
        assertEquals(TokenKind.NUMBER, tokens[0].kind)
        assertEquals("3.14", tokens[0].value)
    }

    @Test
    fun `tokenizes number starting with decimal point`() {
        val tokens = tokenize(".5")
        assertEquals(1, tokens.size)
        assertEquals(TokenKind.NUMBER, tokens[0].kind)
        assertEquals(".5", tokens[0].value)
    }

    @Test
    fun `tokenizes all operators`() {
        val tokens = tokenize("+ - * / % **")
        assertEquals(6, tokens.size)
        assertEquals(TokenKind.PLUS, tokens[0].kind)
        assertEquals("+", tokens[0].value)
        assertEquals(TokenKind.MINUS, tokens[1].kind)
        assertEquals("-", tokens[1].value)
        assertEquals(TokenKind.STAR, tokens[2].kind)
        assertEquals("*", tokens[2].value)
        assertEquals(TokenKind.SLASH, tokens[3].kind)
        assertEquals("/", tokens[3].value)
        assertEquals(TokenKind.PERCENT, tokens[4].kind)
        assertEquals("%", tokens[4].value)
        assertEquals(TokenKind.POWER, tokens[5].kind)
        assertEquals("**", tokens[5].value)
    }

    @Test
    fun `tokenizes parentheses`() {
        val tokens = tokenize("(1)")
        assertEquals(3, tokens.size)
        assertEquals(TokenKind.LPAREN, tokens[0].kind)
        assertEquals(TokenKind.NUMBER, tokens[1].kind)
        assertEquals(TokenKind.RPAREN, tokens[2].kind)
    }

    @Test
    fun `tokenizes complex expression`() {
        val tokens = tokenize("2 + 3 * (4 - 1)")
        assertEquals(9, tokens.size)
        assertEquals(TokenKind.NUMBER, tokens[0].kind)
        assertEquals("2", tokens[0].value)
        assertEquals(TokenKind.PLUS, tokens[1].kind)
        assertEquals(TokenKind.NUMBER, tokens[2].kind)
        assertEquals("3", tokens[2].value)
        assertEquals(TokenKind.STAR, tokens[3].kind)
        assertEquals(TokenKind.LPAREN, tokens[4].kind)
        assertEquals(TokenKind.NUMBER, tokens[5].kind)
        assertEquals("4", tokens[5].value)
        assertEquals(TokenKind.MINUS, tokens[6].kind)
        assertEquals(TokenKind.NUMBER, tokens[7].kind)
        assertEquals("1", tokens[7].value)
        assertEquals(TokenKind.RPAREN, tokens[8].kind)
    }

    @Test
    fun `tokenizes power and star correctly`() {
        val tokens = tokenize("2**3*4")
        assertEquals(5, tokens.size)
        assertEquals(TokenKind.NUMBER, tokens[0].kind)
        assertEquals("2", tokens[0].value)
        assertEquals(TokenKind.POWER, tokens[1].kind)
        assertEquals("**", tokens[1].value)
        assertEquals(TokenKind.NUMBER, tokens[2].kind)
        assertEquals("3", tokens[2].value)
        assertEquals(TokenKind.STAR, tokens[3].kind)
        assertEquals("*", tokens[3].value)
        assertEquals(TokenKind.NUMBER, tokens[4].kind)
        assertEquals("4", tokens[4].value)
    }

    @Test
    fun `tokenizes expression without spaces`() {
        val tokens = tokenize("1+2")
        assertEquals(3, tokens.size)
        assertEquals(TokenKind.NUMBER, tokens[0].kind)
        assertEquals("1", tokens[0].value)
        assertEquals(TokenKind.PLUS, tokens[1].kind)
        assertEquals(TokenKind.NUMBER, tokens[2].kind)
        assertEquals("2", tokens[2].value)
    }

    @Test
    fun `throws error on multiple decimal points`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            tokenize("1.2.3")
        }
        assert(ex.message!!.contains("Unexpected character '.'"))
    }

    @Test
    fun `throws error on unexpected character`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            tokenize("2 @ 3")
        }
        assert(ex.message!!.contains("Unexpected character '@'"))
        assert(ex.message!!.contains("position 2"))
    }
}
